### PR TITLE
Add configuration validation and refactor handling of unknown identifiers

### PR DESCRIPTION
**Pull request by mikabytes**
- Add validation using Ajv and JSON Schema 
- Moved the "unknown identifiers" management there
- Moved the default schema there

Our configuration schema is big and complex. This is mostly due to every property having the ability to also be a function (dynamic configuration). 

Several properties don't belong in configuration. We might want to look into extracting them at a later point. Examples are `cacheLocation`, `coreModules`.. These are also undocumented. They bring in a hairball of dependencies that don't make sense for a schema to deal with. It would clean things up a bit to refactor them out.

This implementation will return messages on bad configuration and delete the offending keys. Alternatively, we could crash, but seeing as we have earlier ignored them this seems to be the most reasonable way to go.

Changed one test using Set to use Array instead, as I couldn't validate Sets using `instanceof`. Not sure why. But it works.

### CODE REVIEW

**Comment by mikabytes** [comment]
@lencioni If it's convenient for you, would you mind having a look at this code? 

Or, if you're pressed for time, perhaps I should go ahead and merge it. Should be fairly safe.

Thanks

File: README.md
Comment (lencioni) [inline README.md:537]: Why remove this?

File: lib/Configuration.js
Comment (lencioni) [inline lib/Configuration.js:30]: This seems to be unused now

File: lib/configurationSchema.js
Comment (lencioni) [inline lib/configurationSchema.js:288]: Should probably change the back ticks to single quotes

File: lib/configurationSchema.js
Comment (lencioni) [inline lib/configurationSchema.js:378]: Do we want this to remove duplicates?

File: lib/configurationSchema.js
Comment (lencioni) [inline lib/configurationSchema.js:386]: Instead of saying "key" I think it might make more sense to say "value at" or something similar. It might also be nice to log typeof v.

File: lib/configurationSchema.js
Comment (lencioni) [inline lib/configurationSchema.js:388]: What do you think about using Object.hasOwn here instead?

File: lib/configurationSchema.js
Comment (lencioni) [inline lib/configurationSchema.js:408]: I appreciate your attention to detail but I do wonder how important is it to preserve this exact behavior?

File: lib/configurationSchema.js
Comment (lencioni) [inline lib/configurationSchema.js:418]: Use quotes here too

File: lib/configurationSchema.js
Comment (lencioni) [inline lib/configurationSchema.js:433]: This could be: .filter(Boolean)

File: lib/configurationSchema.js
Comment (lencioni) [inline lib/configurationSchema.js:427]: Should this use console.error?

**Review by lencioni — APPROVED** [review]

**Comment by lencioni** [comment]
Thanks for doing this! I reviewed on a phone so sorry about any weird formatting

File: lib/Configuration.js
Comment (mikabytes) [reply lib/Configuration.js:30]: Woah, good catch.

I sorted all the configuration options in another PR. This caused a merge conflict here, and I must've done somthing wrong when resolving it.

I have re-added it now.

File: lib/configurationSchema.js
Comment (mikabytes) [reply lib/configurationSchema.js]: Ah, old habits die hard.

In a previous company, we resolved the "single quotes vs double quotes" debate with "always use backticks".

I've changed it to single quotes now.

File: lib/configurationSchema.js
Comment (mikabytes) [reply lib/configurationSchema.js:378]: This function was moved from [configuration.js](https://github.com/Galooshi/import-js/pull/612/files/ae686c28c6cd7d39c63c44368f56caf3f68cdb25#diff-94c307851d4ad36c747a9f883703a4c1f964b82c23320b47899c4e543fd9a8f9L23) because the defaults depend on it.

I would suggest that changes to it be made in another PR, trying to keep this as pure "configuration schema validation" as possible. What do you think?

File: lib/configurationSchema.js
Comment (mikabytes) [reply lib/configurationSchema.js]: Changed it to 
```
`Expected schema key '${k}' to be an object, but it was of type '${typeof v}'. Got: ${v}`
```

This error will only happen when we change the defaults in code, so it should be fairly obvious what key has changed, but it's always nice to provide as much information as possible.

File: lib/configurationSchema.js
Comment (mikabytes) [reply lib/configurationSchema.js:390]: Either one is okay by me. Possibly one downside of using Object.hasOwn is that we would be losing support for Node <16.8. That might be okay, it's a bit over two years old. I'd probably want a higher benefit/disadvantage ratio.

Maybe we better deal with that the day we choose to use `hasOwnProperty` as a key for our configuration :smile:

File: lib/configurationSchema.js
Comment (mikabytes) [reply lib/configurationSchema.js:410]: Not important at all :smile: But it helps reduce the scope of this PR and lowers my anxiety of having accidentally introduced some unwanted side-effect.

I was thinking of improving error messages in an upcoming PR. 

Faulty identifiers / bad configuration would be very nice to have the information of what the offending value was, and how to fix it. Ajv gives me a lot of duplicate error messages, as well as false positives when using `anyOf`, so I think that stuff will need a bit of care anyway.

File: lib/configurationSchema.js
Comment (mikabytes) [reply lib/configurationSchema.js]: Yep

File: lib/configurationSchema.js
Comment (mikabytes) [reply lib/configurationSchema.js]: I would have preferred something that conveys my intention better, "isTruthy" or something, but Boolean will have to do :+1:

File: lib/configurationSchema.js
Comment (mikabytes) [reply lib/configurationSchema.js]: Ah, I left that in there by mistake. I've removed it now.

Logging things in low-level modules makes the tests very chatty, and I think that's anyway not the right place to communicate this information. We already return the error messages, so it is the responsibility of the caller function to do so, if needed.

**Comment by mikabytes** [comment]
@lencioni Thanks for your feedback, much appreciated!

I put some comments of my own. I'll wait for you to review them, and the changes, before merging this.

File: lib/configurationSchema.js
Comment (lencioni) [reply lib/configurationSchema.js:378]: Makes sense

File: lib/configurationSchema.js
Comment (lencioni) [reply lib/configurationSchema.js]: Better but still the wording is a bit off I think. We are talking about the value at that key, not the key itself, but the way it is rded could be understood as referring to the type of the key itself and not its value.

File: lib/configurationSchema.js
Comment (lencioni) [reply lib/configurationSchema.js:390]: Yeah, or could use Object.prototype.hasOwnProperty.call

File: lib/configurationSchema.js
Comment (lencioni) [reply lib/configurationSchema.js:410]: Improving error messages sounds great to me. I wonder if anything could be added to the schema that would help with that, or if ajv offers any options to tweak the output it generates out of the box. Or maybe someone else has put out a package that makes ajv errors really nice?

File: lib/Configuration.js
Comment (lencioni) [reply lib/Configuration.js:30]: I think you responded to the wrong comment here. This array is still unused.

Comment (mikabytes) [reply lib/Configuration.js]: Ah yes. Apologies. I responded to the wrong comment.

The unused variable 'messages' has now been removed.

File: lib/configurationSchema.js
Comment (mikabytes) [reply lib/configurationSchema.js:410]: Good ideas. There must be. I will look into it.

I was also thinking we might move documentation into the JSON schema and auto-generate it. I came across [jsonschema2md](https://github.com/adobe/jsonschema2md) that might be interesting for this. But that's yet another PR.

Comment (mikabytes) [reply lib/configurationSchema.js:410]: Started #615 for invalid configuration error messages and #616 for auto-generating documentation

**Pull request merged**
